src/main/scala/org/allenai/plugins/DeployPlugin.scala

```diff
@@ -420,51 +449,95 @@ object DeployPlugin extends AutoPlugin {
+    * and exit the program if any are missing.
     */
-  def validateAsMap(targetName: String, targetConfig: Config): Map[String, String] = {
-    // Validate that the target has all the required environment variables.
+  def validateEnv(): Unit = {
     val requiredEnv = Seq("AWS_PEM_FILE")
-    for (variable <- requiredEnv) {
-      val value = sys.env.get(variable).getOrElse {
-        throw new IllegalArgumentException(s"Environment variable ${variable} undefined.")
-      }
+    requiredEnv foreach { variable =>
+      val value = sys.env.getOrElse(variable, {
+        throw new IllegalArgumentException(s"Environment variable '$variable' undefined.")
+      })
 
-      require(!value.isEmpty, "Environment variable ${value} is empty.")
+      require(!value.isEmpty, s"Environment variable '$variable' is empty.")
     }
+  }
+
+  /** Container for config needed to deploy a project to a host.
+    * @param host The fully-qualified name of the host the project should be deployed to
+    * @param directory The directory on the host in which the project's code should be placed
+    * @param sshUser The username to use when accessing the host
+    * @param startupScript The path, relative to the deploy directory, to the script that should
+    *                      be used to restart the project after deploying the new code
+    * @param startupArgs Optional arguments to pass to the startup script
+    */
+  case class HostConfig(
+    host: String,
+    directory: String,
+    sshUser: String,
+    startupScript: String,
+    startupArgs: Option[Seq[String]]
+  )
+
+  /** Wrapper for many [[org.allenai.plugins.DeployPlugin.HostConfig]] objects, along with any data
+    * that should be shared between them, when deploying a project.
+    * @param hostConfigs configuration objects specifying how the project should be deployed to
+    *                    many hosts
+    * @param projectVersion the version of the project being deployed
+    */
+  case class DeployConfig(hostConfigs: Seq[HostConfig], projectVersion: Option[String])
+
+  /** Transform the given [[com.typesafe.config.Config]] object into a corresponding
+    * [[org.allenai.plugins.DeployPlugin.DeployConfig]] object.
+    * @param targetName the name of the target to print on error
+    * @param targetConfig env-level deployment config to convert
+    */
+  def parseConfig(targetName: String, targetConfig: Config): DeployConfig = {
 
-    // Validate that the target has all the required keys.
-    val requiredKeys = Seq("deploy.host", "deploy.directory",
-      "deploy.startup_script", "deploy.user.ssh_username")
-    val requiredKeyPairs: Seq[(String, String)] = for {
-      key <- requiredKeys
-    } yield key -> {
-      try {
-        targetConfig.getString(key)
+    DeployConfig(
+      hostConfigs = Seq(parseHostConfig(targetConfig.getConfig("deploy"))),
+      projectVersion = Try(targetConfig.getString("project.version")).toOption
```

Comment:
I don't think that's what you want. Typesafe config shouldn't throw an exception if a key is missing - it should return `null`.

You ought to be able to do `Option(targetConfig.getString("project.version"))`.